### DISCUSSION TITLE
Calendar: Why do hours, minutes and seconds go to 24/60/61 and not 23/59/60?

### DISCUSSION THREAD

**Discussion by Ekleog** [unicode-org/icu4x#3086] [Q&A] [answered]
I'm trying to consume a `DateTime`, and turn it back into a number of seconds since posix epoch.

I'm having questions related to the range of `IsoHour` & co.

The ranges are currently documented as being `[0, 24]`, `[0, 60]` and `[0, 61]`. I do not understand why they contain the last value?

It looks to me like [in early 2021](https://github.com/unicode-org/icu4x/commit/324d9edd2def40e0d000646748172cce370d92db#diff-522db5380f4c987bc355c36a43a6c1ec88fff0828de0c511ad4d0ff86ccea5e5R341), the units were set with an exclusive bound at the end, seeing how "week day" is 7 and one of the rare things everyone seems to agree on is the duration of a week.

However, the comment about the current allowable ranges [seems to have been introduced without explanation](https://github.com/unicode-org/icu4x/commit/bb55be57d6296cce5861453cc7f09b999e95f3b7#diff-8c3f709910d2adfe84ec77d7a1c0ea6c95293086c1ef1a9293b8f8d2cbac6074R253).

Do you know why the allowable ranges are `0-24, 0-60, 0-61`, and not `0-23, 0-59, 0-60` (or `1-24, 1-60, 1-61`)? Are there some calendars that mean different things by the hours / minutes / seconds than ISO 8601, and if so why is this type called `IsoMinutes`? (though some versions of ISO 8601 seem to allow 24:00:00, that wouldn't explain for 60 minutes / 61 seconds)

**Comment by zbraniecki** [comment]
Please, use Discussions for such questions.

**Comment by sffc** [comment, marked as answer]
The extra second is to handle leap seconds, and then all the fields support one greater than the maximum cyclic value due, I believe, to accommodate certain situations that express the difference between "last unit of previous time period" versus "first unit of next time period".

**Reply by Ekleog** [reply]
Got it, thank you! (And sorry for not having used discussions, I've literally never been on a project that actually uses them yet :sweat_smile: )
